### ISSUE TITLE
Stagger parallel uploads

### ISSUE THREAD

**Issue by m-cat** [tus/tus-js-client#443]
**Is your feature request related to a problem? Please describe.**
The problem we ran into at Skynet Labs was that our server was taking a few seconds to process each chunk after uploading. Because parallel chunks would start and end at roughly the same time, there were considerable gaps in the upload.

**Describe the solution you'd like**
We implemented a scheduling algorithm to optimally stagger the chunks. This is off by default and the stagger amount is configurable by the user. We already implemented this in our fork, just let me know if there is interest in a PR. If this feature is too niche, feel free to close this issue.

**Describe alternatives you've considered**
None that we could think of. Our staggering algorithm resulted in significant improvements in total upload times.

**Can you provide help with implementing this feature?**
We already implemented this feature, just checking for interest on your end. No worries if this is too niche -- we personally found it extremely useful, despite the added complexity.

**Additional context**
![image](https://user-images.githubusercontent.com/6035856/183846891-f3b0bbde-4665-4d1d-aef9-129e49dec5a5.png)
The first gap is the biggest because both uploads finish at approximately the same time. But then the gaps become smaller since the uploads shift a bit.

What we did was finetune to start only one upload and once that upload is almost done, it kicks off the parallel one. Then when that one is done it uploads a chunk for the first one again. This resulted in a pretty amazing upload experience.

Thank you for working with us on https://github.com/tus/tus-js-client/pull/299!

**Comment by Acconut** [comment]
Interesting proposal, but I still do not fully understand what it is about? Did you implement a feature which can schedule/trigger the beginning of the upload of different partial parts? I am not sure how this could yield an improvement in upload times. Also, I don't know to make of the included graphic. All in all, this is quite confusing for me right now.

You might also be able to implement this scheduling directly without modifications by returning a Promise in the onBeforeRequest callback and only resolving the Promise if the associated upload should begin.

**Comment by m-cat** [comment]
> Did you implement a feature which can schedule/trigger the beginning of the upload of different partial parts?

Yes, specifically we were scheduling each individual chunk within the parts.

> I am not sure how this could yield an improvement in upload times.

If we had 2+ chunks going in parallel, they would tend to complete at the same time. Our server was then processing for a few seconds after having received the chunks, and during this time no uploading was occurring. 

Also, we found that the throughput was the same whether we had 2+ chunks in parallel or just one at once -- the only benefit of parallel uploads for us was in filling in the gaps in uploading.

> You might also be able to implement this scheduling directly without modifications by returning a Promise in the onBeforeRequest callback and only resolving the Promise if the associated upload should begin.

Interesting, I think that would probably work for most use cases. We needed to optimize it a bit more because, in combination with #299, we had some parts that were different in size and we wanted to schedule the larger parts, first.

I'm thinking this is probably much too niche for our specific use case so I'll go ahead and close it. Thanks for taking the time!
